### PR TITLE
CSM Version Module for CSM Version Agnostic CSI

### PR DESCRIPTION
**Pull request by rustydb**
### Summary and Scope

- Fixes: MTL-2111

#### Issue Type

- RFE Pull Request

This adds functions for handling varying versions of CSM, providing methods for comparing to the current version and finding the version of an installation or upgrade environment. These new functions enable developers to write code for multiple CSM versions, making CSI agnostic to a specific CSM version.

***Inherently this also enables using `csm-version: 1.5` for CSM 1.5 installations and upgrades.***

***GOOD When `csm-version` is above the minimum version, `CSM_RELEASE` is set, and both have matching `<major>.<minor>` version:***

`csi` detects that the environment variable and the configured CSM version match, giving an affirmation to the user that generated files will work for the detected environment. 

```bash
redbull-pit:/var/www/ephemeral/prep # csi config init
2023/03/14 18:19:09 Using config file: /var/www/ephemeral/prep/system_config.yaml
2023/03/14 18:19:09 Interpreting [CSM_RELEASE=1.4.0-beta.64] as v1.4.0-beta.64
2023/03/14 18:19:09 [csm-version] was set to [v1.4]; All inputs are targeted for CSM v1.4
# continues ...
```

***OKAY When `csm-version` is above the minimum version, and `CSM_RELEASE` is unset:***

`csi` warns that no environment variable was set to compare the configured version to. This is okay, `csi` does not require this environment variable but it will be unable to know if the user is generating configurations for the correct environment.

```bash
redbull-pit:/var/www/ephemeral/prep # csi config init
2023/03/14 18:19:40 Using config file: /var/www/ephemeral/prep/system_config.yaml
2023/03/14 18:19:40 CSM installation or upgrade presence is unknown; CSM_RELEASE was not found in the environment
2023/03/14 18:19:40 [csm-version] was set to [v1.4]; All inputs are targeted for CSM v1.4
# continues ...
```

***FATAL when `csm-version` and `CSM_RELEASE` do not match each other's `<major>.<minor>`.***

`csi` exits immediately in this case and alerts the user that their parameters do not align with the detected environment (e.g. generating 1.0 configurations for a 1.4 install).

```bash
redbull-pit:/var/www/ephemeral/prep # csi config init
2023/03/14 17:53:02 Using config file: /var/www/ephemeral/prep/system_config.yaml
2023/03/14 17:53:02 Interpreting [CSM_RELEASE=1.4.0-beta.64] as v1.4.0-beta.64
2023/03/14 17:53:02 Detected a potential mismatch of parameters and CSM environment!

    csm-version=v1.5                 (a.k.a. v1.5)
    CSM_RELEASE=v1.4.0-beta.64       (a.k.a. v1.4)

ERROR: [v1.4 != v1.5]

- Both values must have matching <major>.<minor> versions
- csm-version must be greater than this CSI's minimum allowed CSM version of: [v1.4]

Since these values did not match one another it is possible that other inputs are also wrong.
If inputs from a prior release are being used, please double-check and/or refresh all other inputs and try again.
# fails!
```

***FATAL When `csm-version` does not meet the minimum allowed CSM version CSI and `CSM_RELEASE` is unset:***

`csi` exits immediately when it is incompatible with the target API and alerts the user. The user is using too new of a cray-site-init package for the given parameters, or their parameters are incorrect.

It also warns if CSM_RELEASE was not set.

```bash
redbull-pit:/var/www/ephemeral/prep # csi config init
2023/03/14 18:18:25 Using config file: /var/www/ephemeral/prep/system_config.yaml
2023/03/14 18:18:25 CSM installation or upgrade presence is unknown; CSM_RELEASE was not found in the environment
2023/03/14 18:18:25 [v1.0 < v1.4] this edition of `csi` is only compatible with CSM v1.4 and higher but the given parameters have csm-version set to: v1.0
# fails!
```
### Prerequisites

- [ ] I have included documentation in my PR (or it is not required)
- [x] I tested this on internal system (if yes, please include results or a description of the test)
- [x] I tested this on a vshasta system (if yes, please include results or a description of the test)

### CODE REVIEW

**Comment by rustydb** [comment]
I need to add tests for the `CanonicalVersion` and `DetectedVersion` functions.

**Review by jacobsalmela — APPROVED** [review]
Works as described in my testing.

File: Makefile [hunk 1]
```diff
@@ -20,13 +20,31 @@
...
 
 ifeq ($(VERSION),)
-VERSION := $(shell git describe --tags | tr -s '-' '~' | tr -d '^v')
+export VERSION := $(shell git describe --tags | tr -s '-' '~' | tr -d '^v')
+endif
+
+# By default, if these are not set then set them to match the host.
+ifeq ($(GOOS),)
+OS := $(shell uname)
+export GOOS := $(call lc,$(OS))
```
[the remark this answers is not in the document]
Comment (rustydb) [reply Makefile:40]: Yeah it's not available, this method will work on any OS though. Lately I've been taking a GNU approach to some things.
I got it from here, which mentions the tr approach as well: https://stackoverflow.com/questions/664601/in-gnu-make-how-do-i-convert-a-variable-to-lower-case

**Comment by rustydb** [comment]
I'm double-checking that the untestable `init.go` code still works, and that the MTL-1949 backwards compatibility mod provides the same 1.4 experience.

**Review by heemstra — APPROVED** [review]

**Pull request merged**